Feat/studio manifests cont

snorrees opened the pull request:
### Description

<!--
What changes are introduced?
Why are these changes introduced?
What issue(s) does this solve? (with link, if possible)
-->

### What to review

<!--
What steps should the reviewer take in order to review?
What parts/flows of the application/packages/tooling is affected?
-->

### Testing

<!--
Did you add sufficient testing for this change?
If not, please explain how you tested this change and why it was not
possible/practical for writing an automated test
-->

### Notes for release

<!--
Engineers do not need to worry about the final copy,
but they must provide the docs team with enough context on:

* What changed
* How does one use it (code snippets, etc)
* Are there limitations we should be aware of

If this is PR is a partial implementation of a feature and is not enabled by default or if
this PR does not contain changes that needs mention in the release notes (tooling chores etc),
please call this out explicitly by writing "Part of feature X" or "Not required" in this section.
-->

robinpyon commented on packages/sanity/src/_internal/manifest/extractManifest.ts:
```diff
@@ -0,0 +1,499 @@
+export function extractWorkspace(workspace: Workspace): ManifestWorkspace {
+  const serializedSchema = extractManifestSchemaTypes(workspace.schema)
+
+  return {
+    name: workspace.name,
+    dataset: workspace.dataset,
+    schema: serializedSchema,
+  }
+}
```
Would it be possible to expose workspace `title` and `subtitle` too? (Which may be more meaningful for users when selecting workspaces in _Create_)

robinpyon replied:
Another late addition to the above – I believe we'll also need to include workspace `basePath` (in order to construct preview URLs in _Create_)

Re: custom providers – I don't see the harm in including, provided it's low effort to to so!

snorrees replied:
We will have title, subtitle and basePath on each workspace. 
We will not have auth providers at this point, as the studio supports functions for these.

robinpyon reviewed (commented):
Really nice! 

It always boggles my mind how much needs to be considered here (as exemplified in `transformCommonTypeFields`). The test suite is comprehensive and really helped my understanding here too – thanks!

Will defer to DX here, but some other points (in addition to the comments below):

1. I feel this should be opt-in to start, especially as we've yet to put it through its paces and there'll likely be more movement here (in addition to the point you've mentioned re: increasing build times for _all users_). Touching the default `sanity build` command seems like a pretty big deal.
2. Wondering if this should exist as a standalone command on `sanity schema` (e.g. `sanity schema build-manifest` or similar), especially as there's some inconsistency with CLI arguments between formats. For example, when `format=manifest`:
  - `workspace` isn't respected, manifest generation will always build _all_ workspaces
  - `force-required-fields` is ignored
  - `path` is treated as a directory path (where it's treated as a file path when `format=groq-type-nodes`)

snorrees commented:
> 1. I feel this should be opt-in to start, especially as we've yet to put it through its paces and there'll likely be more movement here (in addition to the point you've mentioned re: increasing build times for _all users_). Touching the default `sanity build` command seems like a pretty big deal.

Top of the agenda for the meeting with Rune. I also feel opt in, on a new command (Ash actually made that, but I deleted it 🙈 ).

> 2. Wondering if this should exist as a standalone command on `sanity schema` (e.g. `sanity schema build-manifest` or similar), especially as there's some inconsistency with CLI arguments between formats. For example, when `format=manifest`:
> 
> * `workspace` isn't respected, manifest generation will always build _all_ workspaces
> * `force-required-fields` is ignored
> * `path` is treated as a directory path (where it's treated as a file path when `format=groq-type-nodes`)

All good arguments in favor of standalone. Bringing it to the meeting.

snorrees commented:
@robinpyon I've changed the code in the following ways:
- reverted all changed to extract schema code
- added `sanity manifest extract` command
- moved all worker code needed for the manifest to new files, so there is no accidental impact on `sanity schema extract`
- renamed files:
    - create-schema.json
    - <schema-hash>.create-schema.json
    
 Pushed a tagged release: gonna update the Create PR to support the new filenames.

robinpyon commented on packages/sanity/src/_internal/cli/actions/schema/extractAction.ts:
I believe all these changes can be reverted, given that manifest creation is now exclusively handled via `sanity manifest extract`? (Currently, you can still build manifest via `sanity schema extract --format manifest`)

snorrees replied:
Good catch!

robinpyon commented on packages/sanity/src/_internal/cli/commands/index.ts:
```diff
@@ -110,6 +111,7 @@ const commands: (CliCommandDefinition | CliCommandGroupDefinition)[] = [
   previewCommand,
   uninstallCommand,
   execCommand,
+  extractManifestCommand,
```
I believe we'll include a separate `CliCommandGroupDefinition` (e.g. extractManifestCommand) here as well

E.g.
```
// packages/sanity/src/_internal/cli/commands/manifest/extractManifestGroup.ts
import {type CliCommandGroupDefinition} from '@sanity/cli'

const extractManifestGroup: CliCommandGroupDefinition = {
  name: 'manifest',
  signature: '[COMMAND]',
  isGroupRoot: true,
  description: 'Extracts the studio configuration as one more JSON manifest files.',
}

export default extractManifestGroup
```

Without this, the manifest command won't appear in the list when running `sanity --help`, and running `sanity manifest` standalone will yield `Error: Command "manifest" is not available outside of a Sanity project context` (even if you are in a project context)

robinpyon reviewed (commented):
Thanks @snorrees!

Left some comments below – just low hanging fruit to do with how `sanity manifest` appears in CLI help and running it standalone

snorrees commented:
@juice49 adding you as explicit reviewer since you know a bit about the history here. Let me know who else needs to know.

snorrees commented:
Any input on why all the build actions are failing is welcome 😅

juice49 commented on packages/sanity/src/_internal/cli/commands/manifest/manifestGroup.ts:
```diff
@@ -0,0 +1,6 @@
+  name: 'manifest',
+  signature: '[COMMAND]',
+  isGroupRoot: true,
+  description: 'Interacts with the studio configuration.',
```
Is this description correct?

snorrees replied:
Yes, but I was a bit unsure about what to put here. Manifest sort of looks at the whole studio config and extracts part of it.

snorrees replied:
For context `schemaGroup` reads: `Interacts with Sanity Studio schema configurations`

juice49 commented on packages/sanity/src/_internal/cli/actions/manifest/extractManifestAction.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  try {
+    await extractManifest(args, context)
```
Could it be a good idea to add a timeout to this operation, at least for initial iterations before we understand how it might perform for various schemas in the wild? We could set a fairly lengthy value (1 or 2 minutes, perhaps) that would at least ensure we don't block folks.

snorrees replied:
Good point, was considering it but was a lazy boi 😅

snorrees replied:
Added a 2 minute timeout.

Also added a line about which env var to set if you run into an extract error - probably useful if folks run into the timeout. Adding 5 seconds to a built is one thing, adding 2 minutes quite another 😓

juice49 reviewed (dismissed):
This looks great, Snorre! Thank you for picking up my work. I've left a couple of small comments, and also bumped this with the Studio team to see if anybody else is able to take a look.

ryanbonial commented:
This looks good to me, but we should have someone with more knowledge about this area also take a look. @ricokahler

snorrees commented:
Updated the "Failure states" section to show differentiation between running the command and running deploy.

snorrees commented:
Known issue: word break is not handled well

![image](https://github.com/user-attachments/assets/7c429a65-09c4-4dae-bd77-1463dce55783)

ricokahler commented on packages/sanity/src/_internal/manifest/extractWorkspaceManifest.ts:
```diff
@@ -0,0 +1,502 @@
+    : {}
+
+  const objectFields: ObjectFields =
+    type.jsonType === 'object' && type.type && INLINE_TYPES.includes(typeName) && isCustomized(type)
```
do you need `INLINE_TYPES.includes` check? i think checking for `type.jsonType === 'object'` along with your `isCustomized` check should be sufficient

snorrees replied:
Object aliases are not allowed to re-define fields (it will throw). Checking for INLINE_TYPES is required, do determine if it is a object-with-fields definition.

ricokahler reviewed (approved):
it looks good! i like the addition of the fail-safe when running this command during deploy. i think this kind of extracted data is really good to be shared in different spots like the graphql command and now the type gen command. That's something for us to discuss related to SDK though

snorrees commented on packages/sanity/src/_internal/cli/actions/manifest/extractManifestAction.ts:
```diff
@@ -0,0 +1,183 @@
+
+/** Escape-hatch env flags to change action behavior */
+const FEATURE_ENABLED_ENV_NAME = 'SANITY_CLI_EXTRACT_MANIFEST_ENABLED'
+const EXTRACT_MANIFEST_DISABLED = process.env[FEATURE_ENABLED_ENV_NAME] === 'false'
```
Beware the inverted naming here (is it too confusing?)

EXTRACT_MANIFEST_**DISABLED**  = process.env[SANITY_CLI_EXTRACT_MANIFEST_**ENABLED**] === 'false'

So when SANITY_CLI_EXTRACT_MANIFEST_ENABLED is anything other than literal string 'false', extract is disabled.

Yeah.... Looking at it now, it IS too confusing. I think I was trying to avoid implied double negations (!!) in the code, but I really just moved that into the naming, which is worse.

Imma change this to:
```suggestion
const EXTRACT_MANIFEST_ENABLED = process.env[FEATURE_ENABLED_ENV_NAME] !== 'false'
```

Then use `!` at the callsite. I think thats much clearer, and not flipping DISABLED and ENABLED in the constant names.

snorrees commented on packages/sanity/src/_internal/cli/actions/manifest/extractManifestAction.ts:
```diff
@@ -0,0 +1,183 @@
+
+const EXTRACT_FAILURE_MESSAGE =
+  'Unable to extract manifest. Certain features like Sanity Create will not work with this studio.\n' +
+  //TODO: replace this link
```
Working out with stakeholders what we want here. Hold on :)

snorrees replied:
Its going to be:

`Couldn't extract manifest file. Sanity Create will not be available for this studio. Disable this message with SANITY_MANIFEST_EXTRACT_ENABLED=false`

(there wont be a article for this any time soon, so we are doing this instead)

snorrees commented on packages/sanity/src/_internal/manifest/extractWorkspaceManifest.ts:
```diff
@@ -0,0 +1,502 @@
+const transformTypeValidationRule: ValidationRuleTransformer = (rule) => {
+  return {
+    ...rule,
+    constraint:
+      'constraint' in rule &&
+      (typeof rule.constraint === 'string'
+        ? rule.constraint.toLowerCase()
+        : retainSerializableProps(rule.constraint)),
```
Serialize everything that can be serialized (unless I know I know up front that I dont want it), has been my mo. Is lets keep ;)

snorrees commented on packages/sanity/src/_internal/cli/actions/deploy/deployAction.ts:
```diff
@@ -101,16 +102,25 @@ export default async function deployStudioAction(
+    await extractManifestSafe(
+      {
+        ...buildArgs,
+        extOptions: {},
+        extraArguments: [],
```
Non of the deploy flags are eligible for extract, so they are omitted. We can passthrough if you think it is useful, but I cant think of a reason atm.

snorrees commented:
All work on this leg of the Create integration is now concluded as far as I see it.

I've re-tested the latest commit with a tagged release (3.57.2-manifests.74) on another project.

## Compatible schema 

```
> sanity manifest extract
✅ Extracted manifest (1578ms)
```
```
> sanity deploy
✅ Checking project info
✅ Clean output folder (19ms)
✅ Build Sanity Studio (10714ms)
✅ Extracted manifest (1681ms)
✅ Verifying local content
✅ Deploying to Sanity.Studio

Success! Studio deployed to <redacted>
```

## Incompatible schema

```
> sanity manifest extract
ℹ️ Couldn't extract manifest file. Sanity Create will not be available for the studio.
Disable this message with SANITY_CLI_EXTRACT_MANIFEST_ENABLED=false

Error: Failed to load configuration file "sanity.config.ts":
hardcoded-error
(...stacktrace)
```

```
> sanity deploy
✅ Checking project info
✅ Clean output folder (19ms)
✅ Build Sanity Studio (10714ms)
ℹ️ Couldn't extract manifest file. Sanity Create will not be available for the studio.
Disable this message with SANITY_CLI_EXTRACT_MANIFEST_ENABLED=false
✅ Verifying local content
✅ Deploying to Sanity.Studio

Success! Studio deployed to <redacted>
```

snorrees commented:
ping @ricokahler @ryanbonial great if ya'll could look over my latest comments and changes, and review.

The idea is to get this into next weeks release, so we can shore up this part of the Create integration. 
(more chapters to come in other PRs)
